fix(rolldown): sync scoping properly in pre_process_ecma_ast

Boshen opened the pull request:
We need to recreate scoping after each transformation step because none of the steps syncs scoping at present, which is a huge waste that we should fix later in the future.

But in the interim, `define` and `inject` plugins now return a `changed` flag for whether AST has been changed, which is used to skip recreating scoping.
Previously it was recreating semantic even if `define` plugin didn't change anything.
And define plugin is always provided by the value `process.env.NODE_ENV: development`.

This gives us a 5 - 10% performance improvement.

I took the opportunity to refactor the code to make it more consistent.

Fixes https://github.com/vitejs/rolldown-vite/issues/446

Boshen commented:
> Merging https://github.com/rolldown/rolldown/pull/6537 will degrade performances by 10.16%

This is not good, let me double check ...

hyf0 commented:
CI failed due to flaky rollup resolve tests. Rerun would solve it.

Boshen commented:
* **#6537** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6537?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6537?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#6541** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6541?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6537).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hyf0 commented:
### Merge activity

* **Oct 15, 9:05 AM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Oct 15, 9:05 AM UTC**: `hyf0` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Oct 15, 9:05 AM UTC**: `hyf0`  removed this pull request from the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Oct 15, 1:53 PM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Oct 15, 1:53 PM UTC**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Oct 15, 2:07 PM UTC**: Merged by the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).

hyf0 commented:
Is this one ok to merge?

Boshen commented:
Not yet, still figuring out the performance regession.

Boshen committed:
fix(rolldown): sync scoping properly in pre_process_ecma_ast (#6537)

We need to recreate scoping after each transformation step because none of the steps syncs scoping.

I took the opportunity to refactor the code to make it more consistent.

Fixes https://github.com/vitejs/rolldown-vite/issues/446
